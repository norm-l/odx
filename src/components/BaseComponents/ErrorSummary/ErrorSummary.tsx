import React, { useRef, useEffect } from 'react';
import PropTypes from 'prop-types';

export default function ErrorSummary(props) {
  const { errors } = props;

  const errorSummaryRef = useRef<any>(null);
  const localizedVal = PCore.getLocaleUtils().getLocaleValue;
  const localeCategory = 'Messages';

  useEffect( () => {
    if(errorSummaryRef && errorSummaryRef?.current){
      errorSummaryRef.current.focus()
    }}, [errorSummaryRef])

  function onClick(e) {
    const ref = e.target.href.indexOf("#") && e.target.href.split("#").pop();
    const target:any = document.getElementById(ref);
    if (!target) return !1;
    target.focus();
    e.preventDefault();
  };


  return (
    <div ref={errorSummaryRef} className='govuk-error-summary' data-module='govuk-error-summary' tabIndex={-1} >
      <div role='alert'>
        <h2 className='govuk-error-summary__title'>There is a problem</h2>
        <div className='govuk-error-summary__body'>
          <ul className='govuk-list govuk-error-summary__list'>
              {errors.map(error => {
                return <li key={error.fieldId}>
                  <a href={`#${error.fieldId}`} onClick={onClick}>{localizedVal(error.message, localeCategory /* ,localeReference */)}</a>
                </li>
              })}
          </ul>
        </div>
      </div>
    </div>
  );
}

ErrorSummary.propTypes = {
  errors: PropTypes.arrayOf(PropTypes.shape({fieldId: PropTypes.string, message: PropTypes.string})),
};
