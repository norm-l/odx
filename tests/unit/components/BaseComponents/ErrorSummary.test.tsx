import React from 'react';
import renderer from 'react-test-renderer';
import ErrorSummary from '../../../../src/components/BaseComponents/ErrorSummary/ErrorSummary';

it('renders correctly', () => {
  const tree = renderer
    .create(<ErrorSummary errors={[]}></ErrorSummary>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('Renders an error summary component with 3 errors listed', () => {
  let a = ErrorSummary.propTypes;

  const tree = renderer
    .create(<ErrorSummary errors={[{message:'Cannot be blank', fieldId:'textInput1'},
                                  {message:'Date cannot be in the past', fieldId:'dateInput1'},
                                  {message:'Date Cannot be in the future', fieldId:'dateInput2'}
                                ]} />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
