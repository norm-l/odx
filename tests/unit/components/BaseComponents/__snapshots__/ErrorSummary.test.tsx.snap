// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders an error summary component with 3 errors listed 1`] = `
<div
  className="govuk-error-summary"
  data-module="govuk-error-summary"
  tabIndex={-1}
>
  <div
    role="alert"
  >
    <h2
      className="govuk-error-summary__title"
    >
      There is a problem
    </h2>
    <div
      className="govuk-error-summary__body"
    >
      <ul
        className="govuk-list govuk-error-summary__list"
      >
        <li>
          <a
            href="#textInput1"
            onClick={[Function]}
          >
            Cannot be blank
          </a>
        </li>
        <li>
          <a
            href="#dateInput1"
            onClick={[Function]}
          >
            Date cannot be in the past
          </a>
        </li>
        <li>
          <a
            href="#dateInput2"
            onClick={[Function]}
          >
            Date Cannot be in the future
          </a>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`renders correctly 1`] = `
<div
  className="govuk-error-summary"
  data-module="govuk-error-summary"
  tabIndex={-1}
>
  <div
    role="alert"
  >
    <h2
      className="govuk-error-summary__title"
    >
      There is a problem
    </h2>
    <div
      className="govuk-error-summary__body"
    >
      <ul
        className="govuk-list govuk-error-summary__list"
      />
    </div>
  </div>
</div>
`;
